mlflow/models/evaluation/default_evaluator.py

```diff
@@ -1484,111 +1481,117 @@ def _get_error_message_missing_columns(self, metric_name, param_names):
-            {joined_error_message}
+    def _raise_exception_for_malformed_metrics(self, malformed_results, eval_df):
+        output_columns = (
+            [] if self.other_output_columns is None else list(self.other_output_columns.columns)
+        )
+        if self.predictions:
+            output_columns.append(self.predictions)
+        elif self.dataset.predictions_name:
+            output_columns.append(self.dataset.predictions_name)
+        else:
+            output_columns.append("predictions")
 
-            Below are the existing column names for the input/output data:
-            Input Columns: {input_columns}
-            Output Columns: {output_columns}
+        input_columns = list(self.X.copy_to_avoid_mutation().columns)
+        if "target" in eval_df:
+            if self.dataset.targets_name:
+                input_columns.append(self.dataset.targets_name)
+            else:
+                input_columns.append("targets")
 
-            To resolve this issue, you may need to specify any required parameters, or if you are
-            missing columns, you may want to map them to an existing column using the following
-            configuration:
-            evaluator_config={{'col_mapping': {{<missing column name>: <existing column name>}}}}"""
-            stripped_message = "\n".join(l.lstrip() for l in full_message.splitlines())
-            raise MlflowException(stripped_message)
+        error_messages = [
+            self._get_error_message_missing_columns(metric_name, param_names)
+            for metric_name, param_names in malformed_results
+        ]
+        joined_error_message = "\n".join(error_messages)
+        full_message = f"""Error: Metric calculation failed for the following metrics:
+        {joined_error_message}
+
+        Below are the existing column names for the input/output data:
+        Input Columns: {input_columns}
+        Output Columns: {output_columns}
+
+        To resolve this issue, you may need to:
+         - specify any required parameters
+         - if you are missing columns, check that there are no circular dependencies among your
+         metrics, and you may want to map them to an existing column using the following
+         configuration:
+        evaluator_config={{'col_mapping': {{<missing column name>: <existing column name>}}}}"""
+        stripped_message = "\n".join(l.lstrip() for l in full_message.splitlines())
+        raise MlflowException(stripped_message)
+
+    def _order_extra_metrics(self, eval_df):
+        remaining_metrics = self.extra_metrics
+
+        while len(remaining_metrics) > 0:
+            pending_metrics = []
+            failed_results = []
+            did_append_metric = False
+            for metric_tuple in remaining_metrics:
+                result = self._get_args_for_metrics(metric_tuple, eval_df)
+                # cannot calculate the metric yet
+                if isinstance(result, tuple):
```

Comment:
As a reader of this code, it's hard to tell why having a tuple return result means we can't calculate the metric. Can we make this clearer, e.g. by returning an object with two fields like "args" and "ready_for_calculation" (bool)